### ISSUE TITLE
Banner notifications

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#1710]
- [x] a) In the Review stage, after an editor has assigned reviewers, there should be a banner notification - “Awaiting Reviews”. This banner should stay active until all assigned reviewers have responded.
- [x] b) In the Review stage, after an editor has asked the author to submit revisions, there should be a banner notification - “Awaiting Author Revisions”. This banner should stay active until the editor accepts the revisions.
- [x] c) In the Review stage, after an editor has been assigned, there should be a banner notification - “Assign Reviewers”. This banner should stay active until at least one reviewer has been assigned.
- [x] d) In the Copyediting stage, immediately after an editor has moved the submission to copyediting, there should be a banner notification - “Assign a copyeditor using the Add link in the Participants list”. Should the Add text be a link? This banner should stay active until at least one copyeditor has been assigned.
- [x] e) In the Copyediting stage, after an editor has assigned a copyeditor, there should be a banner notification - “Awaiting Copyedits”. This banner should stay active until the copyeditor has responded.
- [x] f) In the Production stage, immediately after an editor has sent a submission to Production, there should be a banner notification - "Assign a user to create galleys using the Add link in the Participants list". Should the Add text be linked? This banner should stay active until the assigned user has responded.
- [x] g) In the Production stage, after an editor has assigned a user to create the galley files, there should be a banner notification - “Awaiting Galleys”. This banner should stay active the assigned user has responded.
- [x] h) S. https://github.com/pkp/pkp-lib/issues/1682

**Comment by asmecher** [comment]
@stranack, I've migrated the various notification issues here to keep them organized.

**Comment by bozana** [comment]
@asmecher and @NateWr, is there somewhere in the system an example of those banner notifications for me to take a look?

**Comment by NateWr** [comment]
@bozana: here's how to trigger one:

1. Forward a submission to the review stage and assign a reviewer.
2. Login as reviewer and complete review.
3. Login as `dbarnes` or whoever is assigned editor and then go to review stage.

You should see a notification that reads "All reviews are in and a decision is needed in Review".

https://github.com/pkp/pkp-lib/blob/master/classes/notification/managerDelegate/review/AllReviewsInNotificationManager.inc.php

And then you're into the Notification Manager system which I don't have a handle on. But I think there are a bunch of constants which define notifications, and those are usually the thing to look to find where in the code a notification is actually called. For instance, `NOTIFICATION_TYPE_ALL_REVIEWS_IN`.

**Comment by bozana** [comment]
@stranack, @asmecher or @NateWr, how are the revisions accepted, for the b) above?

**Comment by asmecher** [comment]
> b) In the Review stage, after an editor has asked the author to submit revisions, there should be a banner notification - “Awaiting Author Revisions”. This banner should stay active until the editor accepts the revisions.

I would suggest making it inactive once the author uploads a revision (I suppose any revision will have to do). When that happens, it should be removed.

That leaves the Editor without any notification that new revisions have actually come in, which IIRC is a known problem discussed elsewhere. (Remember anything about this, @stranack?)

**Comment by asmecher** [comment]
> Also, if the editor asks for revisions before a review is finished, should both notifications be presented: "Awaiting Reviews" and "Awaiting Author Revisions"?

Yes, I think that's better than arbitrarily prioritizing tasks. It would indicate to the editor that a reviewer is still, as far as they know, on the job. That might be an editorial oversight, so that would be useful for the editor to see.

**Comment by bozana** [comment]
@asmecher, I am not sure I understand this part correctly: 

> That should be done through a notification. I'd suggest checking whether any notifications are present, and if not, notifying the (Section) Editors that they need to kick things off.
> 
> After that, any important activity should be heralded by an email from a discussion, so a submission getting forgotten will be less likely, I think.

:-\

**Comment by asmecher** [comment]
I mean that d) and f) can be solved the same way, I think: the "update notification" function can check whether any discussions exist in the stage, and if not, then make sure that notifications are present for editors/section editors assigned to the submission telling them that they need to assign and notify someone to perform copyediting/layout work.

If a discussion is already in place, then we can assume that someone is already working on it.

For journals that don't use copyeditors/proofreaders, I suppose we could check that at least one assistant is available to be assigned before creating the notification. If there isn't anyone, don't creat it.

**Comment by bozana** [comment]
How about, as you said, we check if there is any discussion, but also if there is a copyedited file (by file stage id)? 

Also, I solved it now, that the notification from one stage will be removed when the edititor is moving the submission to the next stage. OK so? This way, if journal is not using copyedior, for example, either the editor will upload a file or send the submission to the production, which (both) will then remove the banner. OK so?

**Comment by bozana** [comment]
And what would be the notification text then?

**Comment by asmecher** [comment]
Yes, I think both those ideas make sense.

For the text -- hmm. I think Kevin's proposed text is still OK; it does mention Copyeditor, but since we're on the Copyediting stage, I think that'll make sense implicitly. We could change it to "Assign a user to perform copyediting..." instead of "Assign a copyeditor..." but I think that's clumsy.

**Comment by bozana** [comment]
Aaaa I think I know what you mean, the discussion will be checked just for d) and copyedited file to remove the e) :-) OK, I think now I understand :-) Sorry :-\ THANKS!!!

**Comment by bozana** [comment]
Ah, I've just realized that in OMP there are already NOTIFICATION_TYPE_REVIEW_ROUND_STATUS notifications and that they are working in OMP but not in OJS although the most of the code is in pkp-lib :-O I will have to take further look to figure out what is going there...

**Comment by bozana** [comment]
@asmecher, I think I found the bug, so that we can use the same review round notification as in OMP? That means, they will be: 
"Waiting for reviewers to be selected. " 
"Awaiting responses from reviewers. " 
"Revisions have been requested. ".
Or shall I change those for OMP to be the same as here?
Thanks!!!

**Comment by NateWr** [comment]
Those phrases sound familiar. I think I changed them from the shorter versions to be more descriptive. But if I recall they were geared towards the author workflow (eg - "Waiting for reviewers to be selected" sounds like it's speaking to the author, who's waiting on the editor to do something).

**Comment by bozana** [comment]
OK, I will leave them as they are :-)
Those notifications are for all users, i.e. for editors and authors.
Unfortunately, I am not sure if this old solution is good :-( -- It is easily implemented -- it just looks after the round status from DB and displays it -- but: it could be that an editor requests the revisions while he  is waiting for a (new) review and in that case only one (the last) status will be shown. Hmmm... I think I will leave it as it is and then we can think about it, because in both OJS and OMP it should be the same solution...

**Comment by bozana** [comment]
Aaaa... I think I know how to solve the problem with revisions -- I will handle them differently and independent from review round statuses, just a manager for them...

**Comment by bozana** [comment]
Ah I made a mistake, it doesn't meter if they are treated independently, because the review round status contains REVIEW_ROUND_STATUS_REVISIONS_REQUESTED... thus the problems stays...

**Comment by bozana** [comment]
Thus, the notification "Revisions have been requested." will stay there as long as the submission doesn't move to the next step (and review round status changes). Eventually, we could consider one more status: REVIEW_ROUND_STATUS_REVISIONS_READY?

**Comment by bozana** [comment]
This is the current solution for review revisions notifications:
![reviewrevisionsnotifications](https://cloud.githubusercontent.com/assets/378421/17930664/bc8990de-6a07-11e6-93bd-6cf8e8bed857.png)

I.e. the first is review status, that cannot be removed/changed until it changes, i.e. this notification will stay there, but the new task notification will appear if the revisions are there. At the moment I don't know when i.e. how to delete it (how to check that the editor has considered the revisions), thus it will stay there at the moment. If a new revision file is uploaded, a new i.e. unread task will be created. This way the editor will be notified anew that a new revision is arrived.
Phew... these notifications... :-(

**Comment by bozana** [comment]
PKP-LIB PR: https://github.com/pkp/pkp-lib/pull/1752
OJS PR: https://github.com/pkp/ojs/pull/988
OMP PR: https://github.com/pkp/omp/pull/327
@asmecher, could you review these PRs?

**Comment by asmecher** [comment]
(BTW, if the OMP tests fail on the static pages plugin, don't worry -- I've just committed a submodule update to solve that.)

**Comment by asmecher** [comment]
Reviewed for now, thanks -- let me know when you'd like another look!

**Comment by bozana** [comment]
@asmecher, I made the changes to PKP-LIB and updated the submodule for OJS and OMP. Could you take a look?

**Comment by asmecher** [comment]
All merged, thanks!
